File: rust_bindings/java_ffm/java/src/main/java/com/sayrer/twitter_text/Autolinker.java
package com.sayrer.twitter_text;

import com.sayrer.twitter_text.autolink_h;
import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;

/**
 * Java wrapper for TwitterTextAutolinker that provides automatic resource management.
 *
 * Example usage:
 * <pre>
 * try (Autolinker autolinker = Autolinker.create(true)) {
 *     String result = autolinker.autolink("Check out https://example.com and @user");
 *     System.out.println(result);
 * }
 * </pre>
 */
public final class Autolinker implements AutoCloseable {

    private final MemorySegment handle;
    private boolean closed;

    private Autolinker(MemorySegment handle) {
        this.handle = handle;
    }

    /**
     * Create a new Autolinker instance.
     *
     * @param noFollow whether to add rel="nofollow" to generated links
     * @return a new Autolinker instance
     */
    public static Autolinker create(boolean noFollow) {
        try {
            MemorySegment ptr =
                (MemorySegment) autolink_h.twitter_text_autolinker_new$handle().invoke(noFollow);
            return new Autolinker(ptr);
        } catch (Throwable t) {
            throw new RuntimeException("Failed to create Autolinker", t);
        }
    }

    /**
     * Auto-link all entities (URLs, hashtags, mentions, cashtags) in the text.
     *
     * @param text the text to process
     * @return HTML string with entities converted to links
     */
    public String autolink(String text) {
        checkNotClosed();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment textSegment = arena.allocateFrom(text);
            MemorySegment resultPtr =
                (MemorySegment) autolink_h.twitter_text_autolinker_autolink$handle()
                    .invoke(handle, textSegment);

            String result = resultPtr.getString(0);

            // Free the returned string
            autolink_h.twitter_text_string_free(resultPtr);

            return result;
        } catch (Throwable t) {
            throw new RuntimeException("Failed to autolink text", t);
        }
    }

    /**
     * Auto-link only @username and @username/list references.
     *
     * @param text the text to process
     * @return HTML string with usernames/lists converted to links
     */
    public String autolinkUsernamesAndLists(String text) {
        checkNotClosed();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment textSegment = arena.allocateFrom(text);
            MemorySegment resultPtr =
                (MemorySegment) autolink_h.twitter_text_autolinker_autolink_usernames_and_lists$handle()
                    .invoke(handle, textSegment);

            String result = resultPtr.getString(0);
            autolink_h.twitter_text_string_free(resultPtr);

            return result;
        } catch (Throwable t) {
            throw new RuntimeException("Failed to autolink usernames and lists", t);
        }
    }

    /**
     * Auto-link only #hashtag references.
     *
     * @param text the text to process
     * @return HTML string with hashtags converted to links
     */
    public String autolinkHashtags(String text) {
        checkNotClosed();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment textSegment = arena.allocateFrom(text);
            MemorySegment resultPtr =
                (MemorySegment) autolink_h.twitter_text_autolinker_autolink_hashtags$handle()
                    .invoke(handle, textSegment);

            String result = resultPtr.getString(0);
            autolink_h.twitter_text_string_free(resultPtr);

            return result;
        } catch (Throwable t) {
            throw new RuntimeException("Failed to autolink hashtags", t);
        }
    }

    /**
     * Auto-link only URLs.
     *
     * @param text the text to process
     * @return HTML string with URLs converted to links
     */
    public String autolinkUrls(String text) {
        checkNotClosed();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment textSegment = arena.allocateFrom(text);
            MemorySegment resultPtr =
                (MemorySegment) autolink_h.twitter_text_autolinker_autolink_urls$handle()
                    .invoke(handle, textSegment);

            String result = resultPtr.getString(0);
            autolink_h.twitter_text_string_free(resultPtr);

            return result;
        } catch (Throwable t) {
            throw new RuntimeException("Failed to autolink URLs", t);
        }
    }

    /**
     * Auto-link only $cashtag references.
     *
     * @param text the text to process
     * @return HTML string with cashtags converted to links
     */
    public String autolinkCashtags(String text) {
        checkNotClosed();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment textSegment = arena.allocateFrom(text);
            MemorySegment resultPtr =
                (MemorySegment) autolink_h.twitter_text_autolinker_autolink_cashtags$handle()
                    .invoke(handle, textSegment);

            String result = resultPtr.getString(0);
            autolink_h.twitter_text_string_free(resultPtr);

            return result;
        } catch (Throwable t) {
            throw new RuntimeException("Failed to autolink cashtags", t);
        }
    }

    /**
     * Set the CSS class for URL links.
     *
     * @param urlClass the CSS class to use
     */
    public void setUrlClass(String urlClass) {
        checkNotClosed();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment classSegment = arena.allocateFrom(urlClass);
            autolink_h.twitter_text_autolinker_set_url_class$handle()
                .invoke(handle, classSegment);
        } catch (Throwable t) {
            throw new RuntimeException("Failed to set URL class", t);
        }
    }

    /**
     * Set the target attribute for URL links.
     *
     * @param urlTarget the target attribute (e.g., "_blank")
     */
    public void setUrlTarget(String urlTarget) {
        checkNotClosed();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment targetSegment = arena.allocateFrom(urlTarget);
            autolink_h.twitter_text_autolinker_set_url_target$handle()
                .invoke(handle, targetSegment);
        } catch (Throwable t) {
            throw new RuntimeException("Failed to set URL target", t);
        }
    }

    /**
     * Set the CSS class for hashtag links.
     *
     * @param hashtagClass the CSS class to use
     */
    public void setHashtagClass(String hashtagClass) {
        checkNotClosed();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment classSegment = arena.allocateFrom(hashtagClass);
            autolink_h.twitter_text_autolinker_set_hashtag_class$handle()
                .invoke(handle, classSegment);
        } catch (Throwable t) {
            throw new RuntimeException("Failed to set hashtag class", t);
        }
    }

    /**
     * Set the CSS class for username links.
     *
     * @param usernameClass the CSS class to use
     */
    public void setUsernameClass(String usernameClass) {
        checkNotClosed();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment classSegment = arena.allocateFrom(usernameClass);
            autolink_h.twitter_text_autolinker_set_username_class$handle()
                .invoke(handle, classSegment);
        } catch (Throwable t) {
            throw new RuntimeException("Failed to set username class", t);
        }
    }

    /**
     * Set the CSS class for cashtag links.
     *
     * @param cashtagClass the CSS class to use
     */
    public void setCashtagClass(String cashtagClass) {
        checkNotClosed();
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment classSegment = arena.allocateFrom(cashtagClass);
            autolink_h.twitter_text_autolinker_set_cashtag_class$handle()
                .invoke(handle, classSegment);
        } catch (Throwable t) {
            throw new RuntimeException("Failed to set cashtag class", t);
        }
    }

    private void checkNotClosed() {
        if (closed) {
            throw new IllegalStateException("Autolinker has been closed");
        }
    }

    @Override
    public void close() {
        if (closed) return;
        closed = true;
        try {
            autolink_h.twitter_text_autolinker_free$handle().invoke(handle);
        } catch (Throwable t) {
            throw new RuntimeException("Failed to free Autolinker", t);
        }
    }
}

File: rust/swift-bindings/Sources/TwitterText/Autolinker.swift
import Foundation
import CTwitterText

/// Automatically links URLs, mentions, hashtags, and cashtags in tweet text
public class Autolinker {
    private let handle: OpaquePointer

    /// Create a new autolinker
    /// - Parameter noFollow: Whether to add rel="nofollow" to links
    public init(noFollow: Bool = true) {
        self.handle = twitter_text_autolinker_new(noFollow)!
    }

    deinit {
        twitter_text_autolinker_free(handle)
    }

    /// Autolink all entities (URLs, mentions, hashtags, cashtags) in the text
    public func autolink(_ text: String) -> String {
        guard let resultPtr = twitter_text_autolinker_autolink(handle, text) else {
            return text
        }

        let result = String(cString: resultPtr)
        twitter_text_string_free(resultPtr)
        return result
    }

    /// Autolink only URLs in the text
    public func autolinkURLs(_ text: String) -> String {
        guard let resultPtr = twitter_text_autolinker_autolink_urls(handle, text) else {
            return text
        }

        let result = String(cString: resultPtr)
        twitter_text_string_free(resultPtr)
        return result
    }

    /// Autolink only usernames and lists in the text
    public func autolinkUsernamesAndLists(_ text: String) -> String {
        guard let resultPtr = twitter_text_autolinker_autolink_usernames_and_lists(handle, text) else {
            return text
        }

        let result = String(cString: resultPtr)
        twitter_text_string_free(resultPtr)
        return result
    }

    /// Autolink only hashtags in the text
    public func autolinkHashtags(_ text: String) -> String {
        guard let resultPtr = twitter_text_autolinker_autolink_hashtags(handle, text) else {
            return text
        }

        let result = String(cString: resultPtr)
        twitter_text_string_free(resultPtr)
        return result
    }

    /// Autolink only cashtags in the text
    public func autolinkCashtags(_ text: String) -> String {
        guard let resultPtr = twitter_text_autolinker_autolink_cashtags(handle, text) else {
            return text
        }

        let result = String(cString: resultPtr)
        twitter_text_string_free(resultPtr)
        return result
    }

    // MARK: - Configuration Properties

    /// Set the URL class attribute
    public func setURLClass(_ className: String) {
        twitter_text_autolinker_set_url_class(handle, className)
    }

    /// Set the URL target attribute
    public func setURLTarget(_ target: String) {
        twitter_text_autolinker_set_url_target(handle, target)
    }

    /// Set the symbol tag
    public func setSymbolTag(_ tag: String) {
        twitter_text_autolinker_set_symbol_tag(handle, tag)
    }

    /// Set the text with symbol tag
    public func setTextWithSymbolTag(_ tag: String) {
        twitter_text_autolinker_set_text_with_symbol_tag(handle, tag)
    }

    /// Set the list class attribute
    public func setListClass(_ className: String) {
        twitter_text_autolinker_set_list_class(handle, className)
    }

    /// Set the username class attribute
    public func setUsernameClass(_ className: String) {
        twitter_text_autolinker_set_username_class(handle, className)
    }

    /// Set the hashtag class attribute
    public func setHashtagClass(_ className: String) {
        twitter_text_autolinker_set_hashtag_class(handle, className)
    }

    /// Set the cashtag class attribute
    public func setCashtagClass(_ className: String) {
        twitter_text_autolinker_set_cashtag_class(handle, className)
    }

    /// Set the username URL base
    public func setUsernameURLBase(_ base: String) {
        twitter_text_autolinker_set_username_url_base(handle, base)
    }

    /// Set the list URL base
    public func setListURLBase(_ base: String) {
        twitter_text_autolinker_set_list_url_base(handle, base)
    }

    /// Set the hashtag URL base
    public func setHashtagURLBase(_ base: String) {
        twitter_text_autolinker_set_hashtag_url_base(handle, base)
    }

    /// Set the cashtag URL base
    public func setCashtagURLBase(_ base: String) {
        twitter_text_autolinker_set_cashtag_url_base(handle, base)
    }

    /// Set the invisible tag attributes
    public func setInvisibleTagAttrs(_ attrs: String) {
        twitter_text_autolinker_set_invisible_tag_attrs(handle, attrs)
    }

    /// Set whether to include the @ symbol in username links
    public func setUsernameIncludeSymbol(_ include: Bool) {
        twitter_text_autolinker_set_username_include_symbol(handle, include)
    }

    /// Set whether to add nofollow attribute
    public func setNoFollow(_ noFollow: Bool) {
        twitter_text_autolinker_set_no_follow(handle, noFollow)
    }
}
